src/mlang/backend_ir/bir_interpreter.ml

```diff
@@ -649,20 +641,52 @@ module Make (N : Bir_number.NumberInterface) = struct
           (Pos.unmark (Mir.Error.err_descr_string err));
         ctx
 
-  let evaluate_variable (p : Bir.program) (ctx : ctx)
+  let evaluate_variable (p : Bir.program) (ctx : ctx) (curr_value : var_value)
       (vdef : Bir.variable Mir.variable_def_) : var_value =
     match vdef with
     | Mir.SimpleVar e -> SimpleVar (evaluate_expr ctx p.mir_program e)
     | Mir.TableVar (size, es) ->
         TableVar
           ( size,
-            Array.init size (fun idx ->
-                match es with
-                | IndexGeneric e ->
-                    evaluate_expr
-                      { ctx with ctx_generic_index = Some idx }
-                      p.mir_program e
-                | IndexTable es ->
+            match es with
+            | IndexGeneric (v, e) ->
+                let i =
+                  match Bir.VariableMap.find v ctx.ctx_vars with
+                  | SimpleVar n -> n
+                  | TableVar _ -> assert false
+                  (* should not happen *)
+                  | exception Not_found ->
+                      raise
+                        (RuntimeError
+                           ( ErrorValue
+                               ( "no value found for dynamic index",
+                                 Pos.get_position e ),
+                             ctx ))
```

Comment:
Could you confirm you have checked that the original compiler does not check out of bound accesses?

Reply:
It does check the lower bound (0) but not the upper one. The code below is stricter than this, which I think is fair enough.

Here it checks if the variable index is valuated and crash if not, whereas the old compiler just discards the assign silently, as he would do for negative index.